quickwit/quickwit-ingest/src/ingest_v2/workbench.rs

```diff
@@ -17,22 +17,108 @@
 };
 use quickwit_proto::ingest::{IngestV2Error, RateLimitingCause};
-use quickwit_proto::types::{NodeId, ShardId, SubrequestId};
+use quickwit_proto::types::{NodeId, Position, ShardId, SubrequestId};
+use tokio::sync::Notify;
 use tracing::warn;
 
 use super::router::PersistRequestSummary;
 
+#[derive(Default)]
+struct PublishState {
+    awaiting_publish: HashMap<ShardId, Position>,
+    already_published: HashMap<ShardId, Position>,
+}
+
+/// A helper for tracking the progress of the publish events when running in
+/// `wait_for` commit mode.
+///
+/// Registers a set of shard positions and listens to [`ShardPositionsUpdate`]
+/// events to assert when all the persisted events have been published. To make
+/// sure that no events are missed:
+/// - the tracker should be created before the persist requests are sent
+/// - the `shard_persisted` method should for all successful persist subrequests
+struct PublishTracker {
+    state: Arc<Mutex<PublishState>>,
+    publish_complete: Arc<Notify>,
+    _publish_listen_handle: EventSubscriptionHandle,
+}
+
+impl PublishTracker {
+    fn new(event_tracker: EventBroker) -> Self {
+        let state = Arc::new(Mutex::new(PublishState::default()));
+        let state_clone = state.clone();
+        let publish_complete = Arc::new(Notify::new());
+        let publish_complete_notifier = publish_complete.clone();
+        let _publish_listen_handle =
+            event_tracker.subscribe(move |update: ShardPositionsUpdate| {
+                let mut state_handle = state_clone.lock().unwrap();
+                for (updated_shard_id, updated_position) in &update.updated_shard_positions {
+                    if let Some(shard_position) =
+                        state_handle.awaiting_publish.get(updated_shard_id)
+                    {
+                        if updated_position >= shard_position {
+                            state_handle.awaiting_publish.remove(updated_shard_id);
+                            if state_handle.awaiting_publish.is_empty() {
+                                publish_complete_notifier.notify_one();
+                            }
+                        }
+                    } else {
+                        // Save this position update in case the publish update
+                        // event arrived before the shard persist response. We
+                        // might build a state that tracks irrelevant shards for
+                        // the duration of the query but that should be fine.
+                        state_handle
+                            .already_published
+                            .insert(updated_shard_id.clone(), updated_position.clone());
+                    }
+                }
+            });
+        Self {
+            state,
+            _publish_listen_handle,
+            publish_complete,
+        }
+    }
+
+    fn shard_persisted(&self, shard_id: ShardId, new_position: Position) {
```

Comment:
this method name is not so great. 

Can we rename it?

`set_shard_pesisted_position` maybe?

Reply:
good point